### DISCUSSION TITLE
CSS Anchoring API 👋

### DISCUSSION THREAD

**Discussion by jh3y** [floating-ui/floating-ui#2074] [General]
Hey 👋 

Wanted to start by saying great work with both popper and floating-ui 👏 I read a statistic that [popper is used in almost 4% of websites](https://techblog.wikimedia.org/2020/11/23/web-performance-case-study-wikipedia-page-previews/)!

From #1804 you're likely aware there is a CSS Anchoring API in the works that got recently promoted to "Editors Draft" in the specs: [drafts.csswg.org/css-anchor-1/](https://drafts.csswg.org/css-anchor-1/). I think the fact this spec exists is likely a great testament to the work you've been doing.

We were wondering if you had any thoughts on the spec? Or what do you think would be really important to be included in the spec? Could a CSS interface be useful for popper and floating-ui?

Your insight would be great as someone who's pretty close to the subject. The API is currently available to experiment with in Chrome Canary behind the "Experimental Web Platform Features" flag and I've got a few demos I'd be happy to share if you're interested.

Keep up the great work!

@jh3y ʕ •ᴥ•ʔ

**Comment by FezVrasta** [comment]
Hi, thanks for the nice words, we are happy you like what we built.

We did review the CSS anchor spec internally and identified some aspects that would make it less than ideal to replace parts, or the entirety, of what Floating UI exposes today.

It's been a few months since I reviewed it the last time so maybe it evolved to address them, but some thoughts in random order are:

1. The spec doesn't mention a system to specify a custom bounding box, most of our users rely on this behavior to define where the tooltip should live;
2. It seems to use layout instead of visual viewport, this would make it difficult to, for example, display a combobox on iOS, as when the keyboard is open it will obscure the element;
3. The API feels very verbose, it takes around 20 lines of CSS rules to define a "flip" behavior, something that through Floating UI is achieved with a single `flip()` middleware;
4. It doesn't seem to provide a system to detect if the element is partially or completely hidden/clipped;
5. I couldn't find any example that addresses arrow positioning;
6. There doesn't seem to be any JavaScript interface to interact with it;

We are happy to talk more about it if you'd like to, even directly on Meet.

**Reply by matthew-dean** [reply]
Have you raised any of these concerns with the CSS working group?

**Reply by FezVrasta** [reply]
No, we never heard back from jh3y so I assumed the feedback was either dismissed or addressed. It would have been ideal if the working group included us in the draft phase of the spec rather than asking for feedback when it was almost complete.

**Reply by jh3y** [reply]
hey 🤙 

sorry – the feedback was passed on and think some of, if not all, gets addressed in the comments below 🤞 

I changed roles not long after becoming no longer involved with the CSSWG

think a lot of those points have been or are being tackled to cover most common use cases
as for notes about the amount of CSS required. to me, that's apples and oranges. it would either be 20 lines of CSS that you could copy or defer to classes/attributes, or however many lines of JavaScript that's producing that middleware, but the user is being abstracted away from. in any case, you're likely to lean towards the more performant solution with less overhead to integrate IMO

have some demos i'd like to throw together using the API so will be nice to see what additions have been made 🙏

**Comment by jh3y** [comment]
Hey! 👋 

Thank you so much for the valuable insights and feedback on the current state of the API. It's really great feedback 🙏 
We are taking a little time to review it and explore how and where the API can cater.

Thank you also for the offer to meet with us!

**Comment by atomiks** [comment]
@jh3y I have some questions about the current state of the CSS anchoring API. Ideally, the positioning part of the library could leverage most of the native CSS API (with fallback for older browsers), and sort of pivot to being more of a wrapper around the positioning logic?

How much of the current middleware functionality can be replicated with only CSS?

<img width="337" alt="Screenshot 2023-03-02 at 5 25 25 pm" src="https://user-images.githubusercontent.com/22450188/222348072-151bc3c7-d10d-450e-99ea-72afa11e3fb1.png">

- Is there an equivalent to [shift](https://floating-ui.com/docs/shift), where it keeps the same "placement"/side but merely shifts itself to stay in view? 
- What is the current state of being able to position arrows? Can they read which current `@try` block is used? If `shift()` is possible, can an arrow know the "shifted" amount such that it can appear centered to the anchor element?
- Is it possible to anchor an element to one that's inside an iframe? We got a fair number of issues about that last year until I fixed it - I don't think it's common but it seems to have a use case.
- Is `size()`'s `availableHeight` and `availableWidth` possible?
- `autoPlacement` likely isn't possible since `@position-fallback` works like `flip()`, though this isn't really a major deal since `flip()` is much more common/useful.
- It looks like `hide()` is possible? In Chrome Canary the floating element hides itself when it's out of view at the bottom (although not at the top?).

Others:

- Is it possible to provide a JavaScript API equivalent for [virtual elements](https://floating-ui.com/docs/virtual-elements), e.g. for following the cursor or positioning a context menu without needing to inject a fake anchor node into the DOM to do so?

**Comment by tabatkins** [comment]
Sorry for the delay in this comment.

> Is there an equivalent to [shift](https://floating-ui.com/docs/shift), where it keeps the same "placement"/side but merely shifts itself to stay in view?

Not currently. This was discussed in <https://github.com/w3c/csswg-drafts/issues/7757>, but it's a chunk of additional complexity to figure out exactly how it would work and how to specify it. (Flipping, on the other hand, is pretty trivial.) If this is important, opening an issue in CSSWG with details on how y'all do it would be great!

> What is the current state of being able to position arrows? Can they read which current @try block is used? If shift() is possible, can an arrow know the "shifted" amount such that it can appear centered to the anchor element?

*Currently* it's not possible to know what position set you're in, but the plan is to add an `@container` query that'll let you select on this.

This approach obviously wouldn't let you pipe in the "shift" amount; that would have to be done somewhat differently (and is part of the complexity of the "shift" ability).

> Is it possible to anchor an element to one that's inside an iframe? We got a fair number of issues about that last year until I fixed it - I don't think it's common but it seems to have a use case.

Ooof, not easily. As far as CSS is concerned, the insides of an iframe are completely invisible to the outer page; the iframe is just an opaque rectangle, no different from an `img`.

In *theory* we could do something about this, but it would be pretty novel. I'm not sure if the layout of iframe contents is tied to the rendering timing of the outer page (it's certainly undefined by CSS, but HTML may or may not say something about it); if not, this is even more troublesome.

> Is size()'s availableHeight and availableWidth possible?

It's not in the current spec, but that information is available. Could you describe some use-cases for it (preferably in a CSSWG issue)?

> autoPlacement likely isn't possible since @position-fallback works like flip(), though this isn't really a major deal since flip() is much more common/useful.

Right, not currently available. In theory it's doable - it would rely on the same sort of information that the fallback/overflow calculation does - but it wasn't brought up as a use-case early. If there are use-cases, we can def pursue it in the future, tho.

> It looks like hide() is possible? In Chrome Canary the floating element hides itself when it's out of view at the bottom (although not at the top?).

This actually isn't possible currently, but it's something I believe definitely needs to be done as part of the initial implementation. I haven't written it up yet because there are some details that weren't certain, but it sounds like y'all have made these decisions already, presumably with good reasons. ^_^

If I'm reading this correctly, y'all's implementation lets you choose between hiding when (a) the positioned element is fully outside of the anchor's containing block (or nearest scrollport?), or (b) the anchor is fully scrolled out of its nearest scroller, or (c) both. Plus a margin changing the size of the positioned element or anchor, as appropriate.

Is this right? Do y'all find both of these important? Is there anything else you wish you would or could do in this space?

> Is it possible to provide a JavaScript API equivalent for [virtual elements](https://floating-ui.com/docs/virtual-elements), e.g. for following the cursor or positioning a context menu without needing to inject a fake anchor node into the DOM to do so?

We don't have one yet, but have an open issue for matching the pointer, at least: <https://github.com/w3c/csswg-drafts/issues/8639>. Absolutely possible, there's nothing weird about the timing here. Do y'all have use-cases for stuff beyond the pointer?

(Hm, I suppose the "position relative to an element inside an iframe" is a reasonable use-case.)

**Comment by tabatkins** [comment]
> The spec doesn't mention a system to specify a custom bounding box, most of our users rely on this behavior to define where the tooltip should live;

I opened an issue that I think addresses this a few days ago <https://github.com/w3c/csswg-drafts/issues/8724>, letting you select an element to use for your overflow calculations (rather than using the positioned element's CB). Would this work?

> It seems to use layout instead of visual viewport, this would make it difficult to, for example, display a combobox on iOS, as when the keyboard is open it will obscure the element;

Hm, this comes out of the fact that it just uses the CB of the positioned element, and that's based on the layout viewport (for fixpos and top-layer stuff). But opting a fixpos into the visual viewport is something we've thought about at least some; @bramus would have a little more info about that, probably.

Worst case, the API I mention above for selecting your containment box could be extended to allow this; the syntax would allow it no problem.

> The API feels very verbose, it takes around 20 lines of CSS rules to define a "flip" behavior, something that through Floating UI is achieved with a single flip() middleware;

This is fixed now; specifying `position: absolute; top: anchor(--foo auto);` automatically aligns your top with with the bottom of your anchor, and auto-flips if necessary. (Or `bottom: anchor(--foo auto);` for the reverse; same with left/right.)

> It doesn't seem to provide a system to detect if the element is partially or completely hidden/clipped;
> I couldn't find any example that addresses arrow positioning;
> There doesn't seem to be any JavaScript interface to interact with it;

I believe I addressed all of these in my preceding response. Please let me know if there are any details I missed responding to!

**Comment by bramus** [comment]
> ([#](https://github.com/floating-ui/floating-ui/discussions/2074#discussioncomment-5641827)) But opting a fixpos into the visual viewport is something we've thought about at least some; @bramus would have a little more info about that, probably.

There’s been some discussion going on in https://github.com/w3c/csswg-drafts/issues/7475 about this, where a new type of viewport has been suggested: the “PosFixedViewport” aka “Unzoomed Visual Viewport”. It has the size of the Visual Viewport, but its size is not affected by pinch-zoom – similar to how the Layout Viewport’s size is not affected by pinch-zoom.

In the visual below _(from [this interactive demo](https://flackr.github.io/web-demos/viewport/#behavior=fixed-viewport))_, you can see the Layout Viewport (blue), Visual Viewport (red), and the Unzoomed Visual Viewport (green):

![image](https://user-images.githubusercontent.com/213073/232731017-edb4e161-c7c8-4471-98ad-2acf429ee2d2.png)

The linked-to issue is leaning towards introducing this new type of viewport, but it still undecided on making it the default or doing it via an opt-in _([which I am leaning to](https://github.com/w3c/csswg-drafts/issues/7475#issuecomment-1182982687))_

However, should that new type of viewport land, it doesn’t cover scenarios where the anchor is not `position: fixed` and just happens to be visually just somewhere between the virtual keyboard and the top window edge because the document’s scroll position puts it there.

Preferably – and I think this is what the OP is suggesting – when trying to position the anchored element, the logic should try and take the presence of the virtual keyboard _(or any other interactive widget for that matter)_ into account.

This would play nice not only with full-width virtual keyboards, but also with [split-keyboards](https://support.apple.com/guide/ipad/use-a-split-keyboard-ipad02663f08/ipados) – You wouldn’t want the anchored element to be rendered underneath any of those interactive widgets or parts thereof.

I can see how this would complicate things implementation-wise, though.

**Comment by xiaochengh** [comment]
The spec has evolved quite a bit after these months. Adding to the previous answers:

@FezVrasta 

> The spec doesn't mention a system to specify a custom bounding box, most of our users rely on this behavior to define where the tooltip should live;

We've decided to add a new property for this purpose: https://github.com/w3c/csswg-drafts/issues/8724

It should get into the spec soon.

@atomiks 

> How much of the current middleware functionality can be replicated with only CSS?

- offset: just a calc like `top: calc(anchor(bottom) + 10px)`
- shift: no direct support yet, but it's still doable by making the target element `position:sticky` inside an anchor-positioned container. See demo https://jsfiddle.net/3xytvL5h/
- flip: `anchor(auto)`
- arrow: we'll introduce a CQ for it in the future, as previously answered. This one should be doable, but is technically very complicated in terms of both spec and implementation (because there's style depending on layout), so we don't expect it to land soon
- size: this one is probably the only one fundamentally not doable, because it requries a layout recalc on every scroll change, which makes scrolling janky and is something we explicitly try to avoid; the current `@position-fallback` requires relayout only when switching to a different fallback position, though
- autoPlacement: not currently supported, but should be doable if the following spec issues can be resolved
  - Specify a custom bounding box for position fallback (see https://github.com/w3c/csswg-drafts/issues/8724, will be added to spec soon)
  - Preserve the previously used fallback position when all fallback positions fail (see https://github.com/w3c/csswg-drafts/issues/8200, proposed)
- hide: we are introducing a new property for the same purpose, though not exactly in the same way. See https://github.com/w3c/csswg-drafts/issues/7758#issuecomment-1568911631
- inline: we currently don't support this. When the anchor element is fragmented (due to inline or multicol layout), the [current spec](https://drafts.csswg.org/css-anchor-position-1/#ref-for-target-anchor-element%E2%91%A0) only allows us to choose "axis-aligned bounding rectangle of the fragments' border boxes". But maybe we can augment the `anchor()` function with an additional parameter so that we can anchor to a specific fragment? I filed https://github.com/w3c/csswg-drafts/issues/8895

**Reply by atomiks** [reply]
@xiaochengh thanks!

I can see some great progress has been made, but these are two really important ones I'm currently confused about:

## `shift`

This isn't just for sticky behavior, it's mainly for preventing collision without changing the placement. For example, the following demo fails when it's near the edge of the viewport:

<img width="631" alt="Screenshot 2023-06-01 at 3 35 20 pm" src="https://github.com/floating-ui/floating-ui/assets/22450188/23697b0a-8aa6-474b-9758-8e3fed5bdcb0">


If it overflows on the inline axis of the placement (alignment), does the `@try` fallback logic also kick in? Or does the collision detection only work on the block axis (i.e. axis running along the side)? 

If it works on the inline axis, one workaround would be to fallback to `bottom-start` / `bottom-end` when it overflows, and not be fully dynamic like `shift()`. However, on narrower screens, if it's wide enough, and depending on the anchor location, it can still overflow the screen due to its size, without it. So it's not completely robust.

## Centering

How do you center the floating element relative to its anchor **without** knowing its width and not using `translate: -50%` (since that can cause blurring on low PPI displays, and then collision detection won't work since it's based on layout)? It seems like it can only have a fixed size / you have to knows its size to do this?

**Reply by xiaochengh** [reply]
Regarding `shift`:

> If it overflows on the inline axis of the placement (alignment), does the `@try` fallback logic also kick in?

Yes.

Looks like `shift` is a frequently demanded feature, so I'm thinking if we should add first-class support for it. I'll need to do more research.

---

Regarding centering:

This can be done properly with `justify-content: center`, though not one line of CSS. See this example in the spec: https://drafts4.csswg.org/css-anchor-position-1/#example-fcdb8337

And if the browser doesn't support `abs(x)`, it can be replaced by `max(x, -x)`.

**Reply by xiaochengh** [reply]
Update regarding `shift`:

We explored more use cases with the technique of anchor-positioned container + sticky-positioned content (as in the [demo](https://jsfiddle.net/3xytvL5h/)), and think like it should be strong enough, and should be the right way to go.

However, there's still substantial spec work to do to figure out the interaction between `anchor-scroll` and sticky positioning. Please follow the progress at https://github.com/w3c/csswg-drafts/issues/8448

**Comment by AndrewKGuan** [comment]
Hi Floating-UI experts,
with the information about the Popover API mentioned above, do you think it's enough browser-level support to swap into the library?  Or is it still very tangled to carve out parts of the feature set to take advantage of the API?

**Reply by AndrewKGuan** [reply]
Quick friendly bump on this question.
With more browsers are supporting the popover API, does it make sense to switch part of the library's underlying implementations to use the API?  The positioning can still be done with the existing code via JS.

**Reply by FezVrasta** [reply]
Cc @atomiks

**Reply by atomiks** [reply]
I guess you're referring to the React library here? I'm not sure as I haven't really looked into it deeply yet. But I might soon

**Reply by AndrewKGuan** [reply]
Thanks for you reply, I'm referring to [the browser level API](https://developer.chrome.com/blog/introducing-popover-api) that's already in Chrome and Safari, and soon (this month?) in Firefox.  What do you think?

**Reply by atomiks** [reply]
I mean the `@floating-ui/react` library is the only library that provides interactions where the popover API can be used. Referring to:

> make sense to switch part of the library's underlying implementations to use the API

The vanilla library (`@floating-ui/dom`) is just the JS "polyfill" for the CSS anchor positioning feature.

**Reply by AndrewKGuan** [reply]
Ah I see, that makes sense.
Please keep me posted.

**Reply by atomiks** [reply]
Ok, I'll give it a go maybe this weekend! I'm assuming there are feature detection checks you can use if the browser supports it? `"popovertarget" in HTMLElement` or such (edit: `HTMLElement.prototype.hasOwnProperty("popover")`)? Will try to report back with any limitations I find.

**Reply by AndrewKGuan** [reply]
Thanks again!  For feature detection, it can be done by checking for the presence of the JS interface methods: https://developer.mozilla.org/en-US/docs/Web/API/Popover_API#instance_methods
For example, something like` if(HTMLElement.hidePopover) {use the API} `

And the "standard" way, if there is such a thing, is `HTMLElement.prototype.hasOwnProperty('popover')`
https://open-ui.org/components/popover.research.explainer/#idl-attribute-and-feature-detection

**Reply by AndrewKGuan** [reply]
Hi there, how was the test with the the popover API going? Thanks!

**Reply by AndrewKGuan** [reply]
Hi again experts,
the CSS Anchor Positioning API has landed in Chromium browsers. 
[https://caniuse.com/css-anchor-positioning](url)
Do you think there could be performance benefits for trying to use the API as much as possible?
Please let me hear your thoughts.

**Reply by AndrewKGuan** [reply]
Quick video guide: [https://www.youtube.com/watch?v=ASb9vO3ARHo](url)

**Reply by FezVrasta** [reply]
Thanks for your interest. I'm confident we can leverage the new API to improve the performance of the library and reduce the bundle size. But I'm afraid we will need a wider adoption browser side before this can be implemented. I'll let @atomiks  have the last word though.

**Reply by AndrewKGuan** [reply]
Thanks @FezVrasta, looking forward to @atomiks' thoughts on this.

**Reply by atomiks** [reply]
Hi @AndrewKGuan, I'm sure using it would be super helpful for the most fundamental situations (the placement computation from `computePosition`, including `offset` or `flip` middleware), but adding new branches to check for its support and ensure things continue to work properly in concert is likely going to be cumbersome and inflate bundle size. 

I haven't got too much time lately for this library to deeply look into how the conditional support could work, but people are welcome to contribute PRs if they'd like. I'd be more motivated if it was already supported across the 3 major browser engines, since an unconditional version that uses the CSS anchoring with no fallback branching, but which builds Floating UI's extended features on top, would be usable in production.

**Reply by AndrewKGuan** [reply]
Thanks for your input and that makes sense.  Hopefully the feature will be supported by all major browsers soon.  At that point do you see the need to feature detect for purposes of backward compatibles in older browsers?
